pkg/controller/controller.go

```diff
@@ -1267,6 +1279,65 @@ func (c *M3DBController) getParentCluster(pod *corev1.Pod) (*myspec.M3DBCluster,
 	return cluster, nil
 }
 
+func (c *M3DBController) cleanupDanglingStorage(ctx context.Context, cluster *myspec.M3DBCluster) error {
+	labelSelector := klabels.SelectorFromSet(labels.BaseLabels(cluster))
+
+	claims, err := c.pvcLister.PersistentVolumeClaims(cluster.Namespace).List(labelSelector)
+	if err != nil {
+		return fmt.Errorf("listing claims: %w", err)
+	}
+
+	claimsByName := map[string]*corev1.PersistentVolumeClaim{}
+	for _, claim := range claims {
+		claimsByName[claim.Name] = claim
+	}
+
+	pods, err := c.podLister.Pods(cluster.Namespace).List(labelSelector)
+	if err != nil {
+		return fmt.Errorf("listing pods: %w", err)
+	}
+
+	// remove all used PVCs from the map
+	for _, pod := range pods {
+		for _, volume := range pod.Spec.Volumes {
+			if volume.PersistentVolumeClaim != nil {
+				delete(claimsByName, volume.PersistentVolumeClaim.ClaimName)
+			}
+		}
+	}
+
+	logger := c.logger.With(
+		zap.String("cluster", cluster.Name),
+		zap.String("namespace", cluster.Namespace))
+
+	for _, claim := range claims {
+		isOldClaim := claim.CreationTimestamp.DeepCopy().Add(-24 * time.Hour).After(time.Now())
+
+		logger.Info("Deleting persistent volume",
+			zap.String("persistent_volume", claim.Spec.VolumeName),
+			zap.Bool("is_old_enough", isOldClaim))
+
+		//err = c.kubeClient.CoreV1().PersistentVolumes().Delete(ctx, claim.Spec.VolumeName, metav1.DeleteOptions{})
+		//if err != nil {
+		//	return fmt.Errorf("deleting persistent volume: %w", err)
+		//}
+
+		logger.Info("Deleting persistent volume claim",
```

Comment:
I'm not 100% sure, but I would expect that PVCs need to be deleted before PVs.

Reply:
My concern is that if we delete PVC and then crash, we don't have any references to track back the PV and we are not cleaning it anymore. 

I will check if after deleting PV, PVC does not create a new PV for itself

Reply:
Have you tried running these operations in this order manually? I thought that it will not allow deleting a PV if there's a claim for it, but I don't know for sure. This blog post seems to address this issue too: https://jhooq.com/k8s-delete-pv-pvc/

Reply:
I will test this out

Reply:
I see that we use `reclaimPolicy: Delete` for our storage classes, so PV will get deleted automatically